test/utils.py

```diff
@@ -303,15 +316,114 @@ def get_empty_chw_tensor(self, *, stream_index: int) -> torch.Tensor:
     frames={},  # Automatically loaded from json file
 )
 
-NASA_AUDIO_MP3 = TestContainerFile(
+
+@dataclass
+class TestAudio(TestContainerFile):
+    """Base class for the *audio* streams of a container (potentially a video),
+    or a pure audio file"""
+
+    stream_infos: Dict[int, TestAudioStreamInfo]
+    # stream_index -> list of 2D frame tensors of shape (num_channels, num_samples_in_that_frame)
+    # num_samples_in_that_frame isn't necessarily constant for a given stream.
+    _reference_frames: Dict[int, List[torch.Tensor]] = field(default_factory=dict)
+
+    # Storing each individual frame is too expensive for audio, because there's
+    # a massive overhead in the binary format saved by pytorch. Saving all the
+    # frames in a single file uses 1.6MB while saving all frames in individual
+    # files uses 302MB (yes).
+    # So we store the reference frames in a single file, and load/cache those
+    # when the TestAudio instance is created.
+    def __post_init__(self):
+        super().__post_init__()
+        for stream_index in self.stream_infos:
+            frames_data_path = _get_file_path(
+                f"{self.filename}.stream{stream_index}.all_frames.pt"
+            )
+
+            self._reference_frames[stream_index] = torch.load(
+                frames_data_path, weights_only=True
+            )
+
+    def get_frame_data_by_index(
+        self, idx: int, *, stream_index: Optional[int] = None
+    ) -> torch.Tensor:
+        if stream_index is None:
+            stream_index = self.default_stream_index
+
+        return self._reference_frames[stream_index][idx]
+
+    def get_frame_data_by_range(
+        self,
+        start: int,
+        stop: int,
+        step: int = 1,
+        *,
+        stream_index: Optional[int] = None,
+    ) -> torch.Tensor:
+        tensors = [
+            self.get_frame_data_by_index(i, stream_index=stream_index)
+            for i in range(start, stop, step)
+        ]
+        return torch.cat(tensors, dim=-1)
+
+    def get_frame_index(
+        self, *, pts_seconds: float, stream_index: Optional[int] = None
+    ) -> int:
+        if stream_index is None:
+            stream_index = self.default_stream_index
+
+        if pts_seconds <= self.frames[stream_index][0].pts_seconds:
+            # Special case for e.g. NASA_AUDIO_MP3 whose first frame's pts is
+            # 0.13~, not 0.
+            return 0
+        try:
+            return next(
+                frame_index
+                for (frame_index, frame_info) in self.frames[stream_index].items()
+                if frame_info.pts_seconds
+                <= pts_seconds
+                < frame_info.pts_seconds + frame_info.duration_seconds
+            )
```

Comment:
I have to admit I don't completely understand what's going on here. I get that we have a generator that does a linear walk through the frames, searching for the first frame that meets our conditions, but how does passing that generator to `next()` get us back the one index we want?

Also, it may be better to do this once, in `__post_init__()` and set up a mapping rather than doing it on every call.

Reply:
Unfortunately we can't build a mapping for this because we are mapping contiguous timestamps to integers. I'll add a comment that `bisect` might make things faster if needed (although for such small arrays, the tests run very fast).

`next(it)` just returns the very first entry in `it`, whether `it` is a list, a tuple, a generator, etc. And then we build the generator in such a way that the first frame in the generator is the frame we want. What's going on is exactly the same as the snippet below:

```py
>>> gen = (i for i in range(10) if i > 5)
>>> next(gen)
6
```

The values in `[0, 5]` were never part of `gen`. The first entry in `gen` is `6`.

Reply:
Wow! I never thought about it that way, in that `next()` on a generator is effectively like `lst[0]` on a sequence.